Better Ghost Selection

13spacemen opened the pull request:
## About The Pull Request
Revived my old PR https://github.com/tgstation/tgstation/pull/68901

Replaces the annoying tgui alert popup "Do you want to be X? | Yes | No | Never for this round" that is hard to read and steals window focus, with a nice clean alert in the top right that counts down. If it's the same event/mob they stack with 2x, 3x, etc. It also shows how many candidates/ghosts are signed up.

The poll alerts have screentips too, they countdown and show if you're signed up, how many people are signed up, if you chose "never for this round" (which is cancelable)
## Why It's Good For The Game
![Screenshot 2023-12-13 030302](https://github.com/tgstation/tgstation/assets/46101244/04061a6b-cd9a-4546-9d71-bba6a6b70d87)

Way easier to see what role is available, you get a nice pic of the role and get it's name in big text, you can cancel "never for this round", and you can cancel signing up for a role before the timer is up
## Changelog
:cl:
refactor: Ghost roles now offer ghosts a clickable poll button. Ghosts can select a role, deselect it, alt-click it for "Never For This Round", can cancel "Never", can see the countdown, and can see how many other people are signed up for the role poll.
/:cl:

13spacemen committed:
add SSpolling subsystem, move poll_candidates and related procs to it
add special poll alert handled by SSpolling
poll alert shows number of candidates, time left

13spacemen committed:
removed get_candidates() and unused priority_candidates list
converted get_candidates() uses to SSpolling system
slightly adjusted to_chat message

13spacemen committed:
added role_name_text to many instances of polling

13spacemen committed:
added pic_source to most polls
some very minor edits

san7890 commented:
I haven't reviewed the code yet, does this still respect the preferences that prevent admins/players from ever seeing a ghost notification? The necessity for the admin pref might not exist anymore since they no longer steal focus (getting irked at the focus stealing is why I implemented it) - but we should still consider them

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+
+	var/jumpable = isatom(pic_source) ? pic_source : null
+
+	var/datum/candidate_poll/P = new(role_name_text, question, poll_time, ignore_category, jumpable)
```
get rid of this and all single-letter variables in this proc

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+		var/act_never = ""
+		if(ignore_category)
+			act_never = "<a href='?src=[REF(A)];never=1'>\[Never For This Round]</a>"
+		to_chat(candidate_mob, "<big>[span_boldnotice("Now looking for candidates [role_name_text ? "to play as \an [role_name_text]." : "\"[question]\""] [act_jump] [act_signup] [act_never]")]</big>")
```
```suggestion
		to_chat(candidate_mob, span_boldnotice("<big>Now looking for candidates [role_name_text ? "to play as \an [role_name_text]." : "\"[question]\""] [act_jump] [act_signup] [act_never]</big>"))
```

ideally we should have a `span_big()` or something but for the meantime this will do the exact same thing while having less wack strings

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+	return candidate_list
+
+/datum/controller/subsystem/polling/proc/is_eligible(mob/M, role, check_jobban, the_ignore_category)
+	if(!M.key || !M.client)
```
ditto no single letter variables here

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+
+	return TRUE
+
+/datum/controller/subsystem/polling/proc/polling_finished(datum/candidate_poll/P)
```
no single letter variables

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+		msg += " | Next: [DisplayTimeText(next_poll_to_finish.time_left())] ([length(next_poll_to_finish.signed_up)] candidates)"
+	return ..()
+
+// The datum that describes one instance of candidate polling
+/datum/candidate_poll
```
i think this should be moved to it's own file in the `/datum` folder instead of being in the same file as the SS.

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+	hash = copytext(md5("[question]_[role ? role : "0"]"), 1, 7)
+	return ..()
+
+/datum/candidate_poll/proc/sign_up(mob/M, silent = FALSE)
```
single letter variable

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+		alert_button.update_candidates_number_overlay()
+	return TRUE
+
+/datum/candidate_poll/proc/remove_candidate(mob/M, silent = FALSE)
```
remove this

san7890 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,312 @@
+
+/datum/candidate_poll/proc/trim_candidates()
+	list_clear_nulls(signed_up)
+	for(var/mob in signed_up)
+		var/mob/M = mob
+		if(!M.key || !M.client)
+			signed_up -= M
```
```suggestion
	for(var/mob/player in signed_up) // are we assuming that `signed_up` has nulls in the list? if not, add `as anything`
		if(isnull(player.key) || isnull(player.client))
			signed_up -= player
```

A) we don't need to typecast outside the for loop.
B) as anything saves any perf concerns (unless we might expect to get nulls or something in this list)
C) `isnull()` is more explicit when it comes to typed references like `client` or stuff that defaults to null like `key`. it's also marginally faster.

san7890 reviewed (changes requested):
things that should be cleaned up- mostly nits but a few code considerations

13spacemen committed:
code readability

YakumoChen commented:
Does it still play the sound / flash the tray icon?

jlsnow301 commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+	icon_state = "template"
+	timeout = 30 SECONDS
+	ghost_screentips = TRUE
+	var/show_time_left = FALSE // If true you need to call START_PROCESSING manually
+	var/image/time_left_overlay // The last image showing the time left
+	var/image/signed_up_overlay // image showing that you're signed up
+	var/image/stacks_overlay
+	var/image/candidates_num_overlay
+	var/image/role_name_overlay
+	var/datum/candidate_poll/poll // If set, on Click() it'll register the player as a candidate
```
turn these into autodoc comments please, 
```
/// comment
var/whatever
```

jlsnow301 commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+		time_left_overlay.plane = plane
+		add_overlay(time_left_overlay)
+		qdel(abstract_object)
+	if(poll)
```
early return `isnull(poll)` probably

jlsnow301 commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+	. = ..()
+	if(!.)
+		return
+	if(poll)
```
early return, move `isnull(poll)` into the previous if and remove this

Mothblocks replied:
Early-return

jlsnow301 commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+		owner.abstract_move(target_turf)
+
+/atom/movable/screen/alert/poll_alert/Topic(href, href_list)
+	if(!poll)
```
`isnull(poll)` for clarity

Mothblocks replied:
How is it possible that this is null?

jlsnow301 commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,216 @@
+	return candidate_list
+
+/datum/controller/subsystem/polling/proc/is_eligible(mob/potential_candidate, role, check_jobban, the_ignore_category)
+	if(!potential_candidate.key || !potential_candidate.client)
```
`isnull(x)`

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+			return PROCESS_KILL
+		cut_overlay(time_left_overlay)
+		var/obj/abstract_object = new
+		abstract_object.maptext = "<span style='font-family: \"Small Fonts\"; font-weight: bold; font-size: 8px; color: [(timeleft <= 10 SECONDS) ? "red" : "white"]; -dm-text-outline: 1px black'>[CEILING(timeleft / 10, 1)]</span>"
```
Small Fonts is not allowed to be used as it does not come packaged on modern Windows.

Use the MAPTEXT class.

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+		if(!role_name_overlay)
+			var/only_question = poll.role ? poll.role : "?"
+			var/obj/abstract_object = new
+			abstract_object.maptext = "<span style='font-family: \"Small Fonts\"; text-align: right; font-size: 7px; color: #B3E3FC; -dm-text-outline: 1px black'>[full_capitalize(only_question)]</span>"
```
Same

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+		qdel(abstract_object)
+	if(poll)
+		if(!role_name_overlay)
+			var/only_question = poll.role ? poll.role : "?"
```
```suggestion
			var/only_question = poll.role || "?"
```

jlsnow301 commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+// The datum that describes one instance of candidate polling
+/datum/candidate_poll
+	var/role // The role the poll is for
+	var/question // The question asked to observers
+	var/duration // The duration of the poll
+	var/atom/jump_to_me
+	var/ignoring_category
+	var/list/mob/signed_up // The players who signed up to this poll
+	var/atom/movable/screen/alert/poll_alert/alert_button
+	var/time_started // The world.time at which the poll was created
+	var/finished = FALSE // Whether the polling is finished
+	var/hash // Used to categorize in the alerts system
```
Convert these to autodoc as well please

jlsnow301 commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+	return ..()
+
+/datum/candidate_poll/proc/sign_up(mob/the_candidate, silent = FALSE)
+	if(!istype(the_candidate) || !the_candidate.key || !the_candidate.client)
```
`isnull()`

jlsnow301 commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+	return TRUE
+
+/datum/candidate_poll/proc/remove_candidate(mob/the_candidate, silent = FALSE)
+	if(!istype(the_candidate) || !the_candidate.key || !the_candidate.client)
```
same

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+			var/obj/abstract_object = new
+			abstract_object.maptext = "<span style='font-family: \"Small Fonts\"; text-align: right; font-size: 7px; color: #B3E3FC; -dm-text-outline: 1px black'>[full_capitalize(only_question)]</span>"
+			abstract_object.maptext_width = 128
+			var/matrix/move_matrix = new
+			move_matrix.Translate(-128, 0)
+			abstract_object.transform = move_matrix
```
Why not `abstract_object.transform = abstract_object.transform.Translate(-128, 0)`?

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+			abstract_object.transform = move_matrix
+			role_name_overlay = image(abstract_object)
+			role_name_overlay.plane = plane
+			add_overlay(role_name_overlay)
+			qdel(abstract_object)
```
This seems *extremely* wasteful especially to be doing every process. Why are we going through this much rigamarole instead of keeping an object and settings it maptext? Why bother with images?

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+			add_overlay(role_name_overlay)
+			qdel(abstract_object)
+		var/num_same = 1
+		for(var/datum/candidate_poll/other_poll in SSpolling.currently_polling)
```
`as anything`?

jlsnow301 reviewed (changes requested):
mostly code standard recs. would love it if you wrote a comment above your proc declarations as well

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+		if(poll.ignoring_category)
+			var/selected_never = FALSE
+			if(owner.ckey in GLOB.poll_ignore[poll.ignoring_category])
+				selected_never = TRUE
+			context[SCREENTIP_CONTEXT_ALT_LMB] = "[selected_never ? "Cancel " : ""]Never For This Round"
+		if(poll.jump_to_me && isobserver(owner))
+			context[SCREENTIP_CONTEXT_CTRL_LMB] = "Jump To"
```
One of these should be right-click. Whichever you think is more common.

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+		return
+	if(poll)
+		var/list/modifiers = params2list(params)
+		if((LAZYACCESS(modifiers, ALT_CLICK)) && poll.ignoring_category)
+			set_never_round()
+			return
+		if((LAZYACCESS(modifiers, CTRL_CLICK)) && poll.jump_to_me)
+			jump_to_pic_source()
+			return
```
Prefer AltClick and CtrlClick.

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+	cut_overlay(stacks_overlay)
+	if(stacks <= 1)
+		return
+	var/obj/abstract_object = new
+	abstract_object.maptext = "<span style='font-family: \"Small Fonts\"; font-size: 7px; color: yellow; -dm-text-outline: 1px black'>[stacks]x</span>"
+	var/matrix/move_matrix = new
+	move_matrix.Translate(3, 2)
+	abstract_object.transform = move_matrix
+	stacks_overlay = image(abstract_object)
+	stacks_overlay.plane = plane
+	add_overlay(stacks_overlay)
+	qdel(abstract_object)
```
Wasteful

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -802,6 +804,166 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+	cut_overlay(candidates_num_overlay)
+	if(!length(poll.signed_up))
+		return
+	var/obj/abstract_object = new
+	abstract_object.maptext = "<span style='font-family: \"Small Fonts\"; font-size: 7px; color: aqua; -dm-text-outline: 1px black'>[length(poll.signed_up)]</span>"
+	var/matrix/move_matrix = new
+	move_matrix.Translate(18, 2)
+	abstract_object.transform = move_matrix
+	candidates_num_overlay = image(abstract_object)
+	candidates_num_overlay.plane = plane
+	add_overlay(candidates_num_overlay)
+	qdel(abstract_object)
```
Wasteful

Mothblocks commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,216 @@
+	log_game("Polling candidates [role_name_text ? "for [role_name_text]" : "\"[question]\""] for [poll_time / 10] seconds")
+
+	// Start firing
+	polls_active = TRUE
```
No reason to have this extra parameter if we can derive it from `total_polls > 0`

Mothblocks commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,216 @@
+		question = "Do you want to play as [full_capitalize(role_name_text)]?"
+	if(!question)
+		question = "Do you want to play as a special role?"
+	log_game("Polling candidates [role_name_text ? "for [role_name_text]" : "\"[question]\""] for [poll_time / 10] seconds")
```
```suggestion
	log_game("Polling candidates [role_name_text ? "for [role_name_text]" : "\"[question]\""] for [DisplayTimeText(poll_time)]")
```

Mothblocks commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,216 @@
+	/// Number of polls performed since the start
+	var/total_polls = 0
+	/// The poll that's closest to finishing
+	var/datum/candidate_poll/next_poll_to_finish
```
This doesn't feel like it should be on the heap at all--introduces the possibility of drift. This should be calculated when things request it.

Mothblocks commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,216 @@
+		var/act_never = ""
+		if(ignore_category)
+			act_never = "<a href='?src=[REF(poll_alert_button)];never=1'>\[Never For This Round]</a>"
+		to_chat(candidate_mob, span_boldnotice("<big>Now looking for candidates [role_name_text ? "to play as \an [role_name_text]." : "\"[question]\""] [act_jump] [act_signup] [act_never]</big>"))
```
Is `<big>` what it was before? This is gross

Mothblocks commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,216 @@
+
+	return possible_candidates
+
+/datum/controller/subsystem/polling/proc/poll_ghost_candidates_for_mobs(question, role, check_jobban, poll_time = 30 SECONDS, list/mobs, ignore_category = null, pic_source, role_name_text)
+	var/list/candidate_list = poll_ghost_candidates(question, role, check_jobban, poll_time, ignore_category, pic_source, role_name_text)
+
+	for(var/mob/potential_mob as anything in mobs)
+		if(QDELETED(potential_mob) || !potential_mob.loc)
+			mobs -= potential_mob
+
+	if(!length(mobs))
+		return list()
+
+	return candidate_list
```
I don't understand what this is for, can you explain?

Mothblocks commented on code/controllers/subsystem/polling.dm:
```diff
@@ -0,0 +1,216 @@
+	currently_polling -= finishing_poll
+
+	// Determine which is the next poll closest the completion or "disable" firing if there's none
+	if(!length(currently_polling))
```
Nit:

```suggestion
	if(currently_polling.len > 0)
```

- `length` obscures errors with null
- Not great to use `0` as falsy

Mothblocks commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+	jump_to_me = poll_jumpable
+	signed_up = list()
+	time_started = world.time
+	hash = copytext(md5("[question]_[role ? role : "0"]"), 1, 7)
```
Why are you doing it like this instead of using an incrementing counter?

Mothblocks commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+	var/atom/jump_to_me
+	var/ignoring_category
+	var/list/mob/signed_up // The players who signed up to this poll
+	var/atom/movable/screen/alert/poll_alert/alert_button
```
Do you not need to qdel_null this on Destroy?

Mothblocks commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+	var/role // The role the poll is for
+	var/question // The question asked to observers
+	var/duration // The duration of the poll
+	var/atom/jump_to_me
```
null this out on Destroy

Mothblocks commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+	var/duration // The duration of the poll
+	var/atom/jump_to_me
+	var/ignoring_category
+	var/list/mob/signed_up // The players who signed up to this poll
```
Cut this list on Destroy

Mothblocks commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+		alert_button.update_candidates_number_overlay()
+	return TRUE
+
+/datum/candidate_poll/proc/never_for_this_round(mob/the_candidate, undoing = FALSE)
```
There's no reason this should handle both putting in and undoing. Make them two separate procs.

Mothblocks commented on code/datums/candidate_poll.dm:
```diff
@@ -0,0 +1,94 @@
+
+/datum/candidate_poll/proc/trim_candidates()
+	list_clear_nulls(signed_up)
+	for(var/mob/the_candidate as anything in signed_up)
```
Remove `the_`

Mothblocks commented on code/game/atom/_atom.dm:
```diff
@@ -861,7 +863,7 @@
 	var/extra_context = ""
 
-	if (isliving(user) || isovermind(user) || isaicamera(user))
+	if(isliving(user) || isovermind(user) || isaicamera(user) || (isobserver(user) && ghost_screentips))
```
Check `ghost_screentips` first. This is one of the most expensive procs in the entire game and so putting something expensive like `isobserver` first (O(n)) will have noticeable losses compared to the simple bool check.

13spacemen committed:
fixed incorrect countdown when stacking poll alerts
improved code readability and responsiveness
ghost_screentips comes first in AND check
added autodocs
used MAPTEXT define

13spacemen committed:
more code improvements, removed unnecessary code

13spacemen committed:
minor improvements

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -40,6 +40,8 @@
 				return thealert
 			// Reset timeout of existing alert
 			var/timeout = initial(thealert.timeout)
+			if(timeout_override)
+				timeout = timeout_override
```
```suggestion
			var/timeout = timeout_override || initial(thealert.timeout)
```

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -809,18 +811,42 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
+	add_overlay(role_overlay)
+
+/atom/movable/screen/alert/poll_alert/Destroy()
+	. = ..()
```
Move to the end and `return ..()`

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -845,63 +871,39 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
-		time_left_overlay = image(abstract_object)
-		time_left_overlay.plane = plane
+		time_left_overlay = new
+		time_left_overlay.maptext = MAPTEXT("<span style='font-weight: bold; color: [(timeleft <= 10 SECONDS) ? "red" : "white"]; -dm-text-outline: 1px black'>[CEILING(timeleft / 10, 1)]</span>")
```
You don't need the text outline, MAPTEXT does that for you. Also use `b` instead of changing font style.

Mothblocks replied:
Double check tho

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -845,63 +871,39 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
-		time_left_overlay = image(abstract_object)
-		time_left_overlay.plane = plane
+		time_left_overlay = new
+		time_left_overlay.maptext = MAPTEXT("<span style='font-weight: bold; color: [(timeleft <= 10 SECONDS) ? "red" : "white"]; -dm-text-outline: 1px black'>[CEILING(timeleft / 10, 1)]</span>")
```
```suggestion
		time_left_overlay.maptext = MAPTEXT("<span style='font-weight: bold; color: [(timeleft <= 10 SECONDS) ? "red" : "white"]; -dm-text-outline: 1px black'>[CEILING(timeleft / (1 SECONDS), 1)]</span>")
```

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -845,63 +871,39 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
-			// Add a small overlay to indicate we've signed up
-			update_signed_up_alert()
+	var/list/modifiers = params2list(params)
+	if((LAZYACCESS(modifiers, ALT_CLICK)) && poll.ignoring_category)
```
Don't need to wrap LAZYACCESS in parentheses

Mothblocks commented on code/_onclick/hud/alert.dm:
```diff
@@ -940,29 +934,24 @@ or shoot a gun to move around via Newton's 3rd Law of Motion."
-	candidates_num_overlay = image(abstract_object)
-	candidates_num_overlay.plane = plane
+	candidates_num_overlay = new
+	candidates_num_overlay.maptext = MAPTEXT("<span style='text-align: right; color: aqua; -dm-text-outline: 1px black'>[length(poll.signed_up)]</span>")
```
Don't need outline

13spacemen committed:
removed unnecessary code in spans, minor cleanups

Admiral-Gigaton commented:
blud this PR is incredibly glitchy man
